### PR TITLE
[#2402] Use the messaging system configured on the tenant

### PR DESCRIPTION
**Pull request by b-abel**
This PR belongs to #2402. It adds support for selecting the messaging system at the tenant level. 
Protocol adapters start AMQP-based event senders *and* Kafka-based event senders (if the configuration for both is provided). At the tenant, a new property is added to select which messaging to use.

### CODE REVIEW

**Comment by b-abel** [comment]
I only added the change for events. When we agree on the strategy, I'll add telemetry.

File: adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java [hunk 1]
```diff
@@ -1031,8 +1077,8 @@ public void registerReadinessChecks(final HealthCheckHandler handler) {
         if (telemetrySender instanceof ServiceClient) {
             ((ServiceClient) telemetrySender).registerReadinessChecks(handler);
         }
-        if (eventSender instanceof ServiceClient) {
-            ((ServiceClient) eventSender).registerReadinessChecks(handler);
+        if (amqpEventSender instanceof ServiceClient) {
+            ((ServiceClient) amqpEventSender).registerReadinessChecks(handler);
```
[the remark this answers is not in the document]
Comment (b-abel) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java:1081]: Nothing obvious as the connection is managed internally in the Kafka client.

[the remark this answers is not in the document]
Comment (sophokles73) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java:1081]: Can we use the Kafka client API to determine if the client is currently connected to the cluster?

[the remark this answers is not in the document]
Comment (b-abel) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java:1077]: There is no way to ask the client that directly. There are a couple of workarounds I found like setting up a separate client that checks if it can connect to the cluster and access a topic, or checking the Prometheus metrics of the cluster. Quarkus implements such checks but disables them by default (https://quarkus.io/guides/kafka#kafka-health-check, https://github.com/smallrye/smallrye-reactive-messaging/blob/3518618b608717d07310fc2e58865ac19abef105/smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/health/BaseHealth.java#L34). 

I think the Kafka way is to try to produce records and if this fails to inspect the exception. We could think about failing the Kubernetes readiness check if producing repeatedly fails with the same connection error (consumers probably need different checks than producers). But I would like to collect more practical experience with our Kafka implementation before working on that topic. I could create an issue for it as the question exceeds the scope of this PR.

[the remark this answers is not in the document]
Comment (sophokles73) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java:1077]: I guess a TODO comment in the code should be enough ...

File: adapter-base/src/main/java/org/eclipse/hono/adapter/MultiMessagingSenders.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import org.eclipse.hono.util.TenantObject;
+
+import io.vertx.core.json.JsonObject;
+
+/**
+ * Encapsulates the Kafka-based and the AMQP-based sender for a message type.
+ *
+ * @param <T> The sender type (intended to be one of {@link org.eclipse.hono.adapter.client.telemetry.EventSender},
+ *            {@link org.eclipse.hono.adapter.client.telemetry.TelemetrySender}, or
+ *            {@link org.eclipse.hono.adapter.client.command.CommandResponseSender}).
```
Comment (sophokles73) [inline adapter-base/src/main/java/org/eclipse/hono/adapter/MultiMessagingSenders.java:29]: these particular types seem to be unrelated to this class, thus I'd rather not introduce a dependency by means of this JavaDoc comment.

File: adapter-base/src/main/java/org/eclipse/hono/adapter/MultiMessagingSenders.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+     * @return true if both senders are {code null}.
+     */
+    public boolean isUnconfigured() {
+        return amqpSender == null && kafkaSender == null;
+    }
+
+    /**
+     * Gets the client to be used for sending messages downstream.
+     *
+     * @param tenant The tenant for which to send messages.
```
Comment (sophokles73) [inline adapter-base/src/main/java/org/eclipse/hono/adapter/MultiMessagingSenders.java:86]: FMPOV we also need to document here which properties of the tenant will determine the returned type of sender

File: adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java [hunk 4]
```diff
@@ -1071,8 +1087,8 @@ public void registerLivenessChecks(final HealthCheckHandler handler) {
         if (commandRouterClient instanceof ServiceClient) {
             ((ServiceClient) commandRouterClient).registerLivenessChecks(handler);
         }
-        if (telemetrySender instanceof ServiceClient) {
-            ((ServiceClient) telemetrySender).registerLivenessChecks(handler);
+        if (telemetrySenders.getAmqpSender() instanceof ServiceClient) {
+            ((ServiceClient) telemetrySenders.getAmqpSender()).registerLivenessChecks(handler);
```
Comment (sophokles73) [inline adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java:1091]: IMHO we should encapsulate these checks and the setting of the liveness/readiness checks within the MultiMessagingSenders class ...

Comment (b-abel) [reply adapter-base/src/main/java/org/eclipse/hono/adapter/AbstractProtocolAdapterBase.java]: I added this in my last commit. Please have a look.

**Review by sophokles73 — APPROVED** [review]
LGTM

**Comment by sophokles73** [comment]
@b-abel Last thing missing is documentation of the extension property, i.e. how can a user configure the messaging type being used.
